LinuxMemoryMeasurementTestFixture.TestPublishMetricsMessage test is flaky

## Description

LinuxMemoryMeasurementTestFixture.TestPublishMetricsMessage is flaky

## Expected Behavior

The test should always succeed.

## Actual Behavior

The test sometimes fail because it relies on having messages processed "fast enough". This has been reported in an ASAN build, most likely due to the fact that ASAN builds are slower than normal builds.

https://github.com/ros-tooling/system_metrics_collector/runs/470335108?check_suite_focus=true

<details>
<summary>Failed test output</summary>

```
2020-02-26T17:21:35.8557391Z test 4
2020-02-26T17:21:35.8557460Z       Start  4: test_linux_memory_measurement_node
2020-02-26T17:21:35.8557524Z 
2020-02-26T17:21:35.8558472Z 4: Test command: /usr/bin/python3 "-u" "/home/runner/work/system_metrics_collector/system_metrics_collector/ros_ws/install/ament_cmake_test/share/ament_cmake_test/cmake/run_test.py" "/home/runner/work/system_metrics_collector/system_metrics_collector/ros_ws/build/system_metrics_collector/test_results/system_metrics_collector/test_linux_memory_measurement_node.gtest.xml" "--package-name" "system_metrics_collector" "--output-file" "/home/runner/work/system_metrics_collector/system_metrics_collector/ros_ws/build/system_metrics_collector/ament_cmake_gtest/test_linux_memory_measurement_node.txt" "--command" "/home/runner/work/system_metrics_collector/system_metrics_collector/ros_ws/build/system_metrics_collector/test_linux_memory_measurement_node" "--gtest_output=xml:/home/runner/work/system_metrics_collector/system_metrics_collector/ros_ws/build/system_metrics_collector/test_results/system_metrics_collector/test_linux_memory_measurement_node.gtest.xml"
2020-02-26T17:21:35.8558649Z 4: Test timeout computed to be: 60
2020-02-26T17:21:35.8558978Z 4: -- run_test.py: invoking following command in '/home/runner/work/system_metrics_collector/system_metrics_collector/ros_ws/build/system_metrics_collector':
2020-02-26T17:21:35.8559465Z 4:  - /home/runner/work/system_metrics_collector/system_metrics_collector/ros_ws/build/system_metrics_collector/test_linux_memory_measurement_node --gtest_output=xml:/home/runner/work/system_metrics_collector/system_metrics_collector/ros_ws/build/system_metrics_collector/test_results/system_metrics_collector/test_linux_memory_measurement_node.gtest.xml
2020-02-26T17:21:35.8559602Z 4: Running main() from /home/runner/work/system_metrics_collector/system_metrics_collector/ros_ws/install/gtest_vendor/src/gtest_vendor/src/gtest_main.cc
2020-02-26T17:21:35.8559712Z 4: [==========] Running 3 tests from 2 test cases.
2020-02-26T17:21:35.8559945Z 4: [----------] Global test environment set-up.
2020-02-26T17:21:35.8560246Z 4: [----------] 1 test from LinuxMemoryMeasurementTest
2020-02-26T17:21:35.8560353Z 4: [ RUN      ] LinuxMemoryMeasurementTest.TestReadInvalidFile
2020-02-26T17:21:35.8560456Z 4: [ERROR] [1582737685.733303781] [ReadFileToString]: unable to parse file: this_will_fail.txt
2020-02-26T17:21:35.8560813Z 4: [       OK ] LinuxMemoryMeasurementTest.TestReadInvalidFile (0 ms)
2020-02-26T17:21:35.8561153Z 4: [----------] 1 test from LinuxMemoryMeasurementTest (0 ms total)
2020-02-26T17:21:35.8561247Z 4: 
2020-02-26T17:21:35.8561484Z 4: [----------] 2 tests from LinuxMemoryMeasurementTestFixture
2020-02-26T17:21:35.8561583Z 4: [ RUN      ] LinuxMemoryMeasurementTestFixture.testManualMeasurement
2020-02-26T17:21:35.8561776Z 4: [       OK ] LinuxMemoryMeasurementTestFixture.testManualMeasurement (19 ms)
2020-02-26T17:21:35.8561862Z 4: [ RUN      ] LinuxMemoryMeasurementTestFixture.TestPublishMetricsMessage
2020-02-26T17:21:35.8561985Z 4: /home/runner/work/system_metrics_collector/system_metrics_collector/ros_ws/src/system_metrics_collector/system_metrics_collector/test/system_metrics_collector/test_linux_memory_measurement.cpp:275: Failure
2020-02-26T17:21:35.8562092Z 4: Expected equality of these values:
2020-02-26T17:21:35.8562180Z 4:   expected_stat.at(stat.data_type)
2020-02-26T17:21:35.8562270Z 4:     Which is: 44.148414611816406
2020-02-26T17:21:35.8562355Z 4:   stat.data
2020-02-26T17:21:35.8562421Z 4:     Which is: 84.423248291015625
2020-02-26T17:21:35.8562524Z 4: /home/runner/work/system_metrics_collector/system_metrics_collector/ros_ws/src/system_metrics_collector/system_metrics_collector/test/system_metrics_collector/test_linux_memory_measurement.cpp:275: Failure
2020-02-26T17:21:35.8562634Z 4: Expected equality of these values:
2020-02-26T17:21:35.8562722Z 4:   expected_stat.at(stat.data_type)
2020-02-26T17:21:35.8562806Z 4:     Which is: 44.148414611816406
2020-02-26T17:21:35.8562891Z 4:   stat.data
2020-02-26T17:21:35.8562957Z 4:     Which is: 98.533164978027344
2020-02-26T17:21:35.8563068Z 4: /home/runner/work/system_metrics_collector/system_metrics_collector/ros_ws/src/system_metrics_collector/system_metrics_collector/test/system_metrics_collector/test_linux_memory_measurement.cpp:275: Failure
2020-02-26T17:21:35.8563170Z 4: Expected equality of these values:
2020-02-26T17:21:35.8563254Z 4:   expected_stat.at(stat.data_type)
2020-02-26T17:21:35.8563337Z 4:     Which is: 44.148414611816406
2020-02-26T17:21:35.8563422Z 4:   stat.data
2020-02-26T17:21:35.8563487Z 4:     Which is: 70.313331604003906
2020-02-26T17:21:35.8563591Z 4: /home/runner/work/system_metrics_collector/system_metrics_collector/ros_ws/src/system_metrics_collector/system_metrics_collector/test/system_metrics_collector/test_linux_memory_measurement.cpp:275: Failure
2020-02-26T17:21:35.8563695Z 4: Expected equality of these values:
2020-02-26T17:21:35.8563778Z 4:   expected_stat.at(stat.data_type)
2020-02-26T17:21:35.8563861Z 4:     Which is: 1
2020-02-26T17:21:35.8563928Z 4:   stat.data
2020-02-26T17:21:35.8564011Z 4:     Which is: 2
2020-02-26T17:21:35.8564113Z 4: /home/runner/work/system_metrics_collector/system_metrics_collector/ros_ws/src/system_metrics_collector/system_metrics_collector/test/system_metrics_collector/test_linux_memory_measurement.cpp:275: Failure
2020-02-26T17:21:35.8564214Z 4: Expected equality of these values:
2020-02-26T17:21:35.8564297Z 4:   expected_stat.at(stat.data_type)
2020-02-26T17:21:35.8564379Z 4:     Which is: 0
2020-02-26T17:21:35.8564444Z 4:   stat.data
2020-02-26T17:21:35.8564524Z 4:     Which is: 14.109916687011719
2020-02-26T17:21:35.8564626Z 4: /home/runner/work/system_metrics_collector/system_metrics_collector/ros_ws/src/system_metrics_collector/system_metrics_collector/test/system_metrics_collector/test_linux_memory_measurement.cpp:275: Failure
2020-02-26T17:21:35.8564733Z 4: Expected equality of these values:
2020-02-26T17:21:35.8564815Z 4:   expected_stat.at(stat.data_type)
2020-02-26T17:21:35.8564900Z 4:     Which is: 84.423248291015625
2020-02-26T17:21:35.8564967Z 4:   stat.data
2020-02-26T17:21:35.8565046Z 4:     Which is: 70.385238647460938
2020-02-26T17:21:35.8565201Z 4: /home/runner/work/system_metrics_collector/system_metrics_collector/ros_ws/src/system_metrics_collector/system_metrics_collector/test/system_metrics_collector/test_linux_memory_measurement.cpp:275: Failure
2020-02-26T17:21:35.8565314Z 4: Expected equality of these values:
2020-02-26T17:21:35.8565397Z 4:   expected_stat.at(stat.data_type)
2020-02-26T17:21:35.8565466Z 4:     Which is: 98.533164978027344
2020-02-26T17:21:35.8565550Z 4:   stat.data
2020-02-26T17:21:35.8565630Z 4:     Which is: 70.385238647460938
2020-02-26T17:21:35.8565733Z 4: /home/runner/work/system_metrics_collector/system_metrics_collector/ros_ws/src/system_metrics_collector/system_metrics_collector/test/system_metrics_collector/test_linux_memory_measurement.cpp:275: Failure
2020-02-26T17:21:35.8565877Z 4: Expected equality of these values:
2020-02-26T17:21:35.8565962Z 4:   expected_stat.at(stat.data_type)
2020-02-26T17:21:35.8566028Z 4:     Which is: 70.313331604003906
2020-02-26T17:21:35.8566114Z 4:   stat.data
2020-02-26T17:21:35.8566195Z 4:     Which is: 70.385238647460938
2020-02-26T17:21:35.8566299Z 4: /home/runner/work/system_metrics_collector/system_metrics_collector/ros_ws/src/system_metrics_collector/system_metrics_collector/test/system_metrics_collector/test_linux_memory_measurement.cpp:275: Failure
2020-02-26T17:21:35.8566398Z 4: Expected equality of these values:
2020-02-26T17:21:35.8566466Z 4:   expected_stat.at(stat.data_type)
2020-02-26T17:21:35.8566550Z 4:     Which is: 2
2020-02-26T17:21:35.8566632Z 4:   stat.data
2020-02-26T17:21:35.8566713Z 4:     Which is: 1
2020-02-26T17:21:35.8566812Z 4: /home/runner/work/system_metrics_collector/system_metrics_collector/ros_ws/src/system_metrics_collector/system_metrics_collector/test/system_metrics_collector/test_linux_memory_measurement.cpp:275: Failure
2020-02-26T17:21:35.8566918Z 4: Expected equality of these values:
2020-02-26T17:21:35.8566985Z 4:   expected_stat.at(stat.data_type)
2020-02-26T17:21:35.8567069Z 4:     Which is: 14.109916687011719
2020-02-26T17:21:35.8567154Z 4:   stat.data
2020-02-26T17:21:35.8567234Z 4:     Which is: 0
2020-02-26T17:21:35.8567335Z 4: /home/runner/work/system_metrics_collector/system_metrics_collector/ros_ws/src/system_metrics_collector/system_metrics_collector/test/system_metrics_collector/test_linux_memory_measurement.cpp:331: Failure
2020-02-26T17:21:35.8567419Z 4: Expected equality of these values:
2020-02-26T17:21:35.8567502Z 4:   3
2020-02-26T17:21:35.8568507Z 4:   test_receive_measurements->GetNumReceived()
2020-02-26T17:21:35.8568605Z 4:     Which is: 2
2020-02-26T17:21:35.8568709Z 4: /home/runner/work/system_metrics_collector/system_metrics_collector/ros_ws/src/system_metrics_collector/system_metrics_collector/test/system_metrics_collector/test_linux_memory_measurement.cpp:352: Failure
2020-02-26T17:21:35.8568818Z 4: Expected equality of these values:
2020-02-26T17:21:35.8568885Z 4:   3
2020-02-26T17:21:35.8569119Z 4:   test_receive_measurements->GetNumReceived()
2020-02-26T17:21:35.8569207Z 4:     Which is: 2
2020-02-26T17:21:35.8569312Z 4: /home/runner/work/system_metrics_collector/system_metrics_collector/ros_ws/src/system_metrics_collector/system_metrics_collector/test/system_metrics_collector/test_linux_memory_measurement.cpp:275: Failure
2020-02-26T17:21:35.8569412Z 4: Expected equality of these values:
2020-02-26T17:21:35.8569496Z 4:   expected_stat.at(stat.data_type)
2020-02-26T17:21:35.8569564Z 4:     Which is: 70.385238647460938
2020-02-26T17:21:35.8569647Z 4:   stat.data
2020-02-26T17:21:35.8569727Z 4:     Which is: 70.330429077148438
2020-02-26T17:21:35.8569828Z 4: /home/runner/work/system_metrics_collector/system_metrics_collector/ros_ws/src/system_metrics_collector/system_metrics_collector/test/system_metrics_collector/test_linux_memory_measurement.cpp:275: Failure
2020-02-26T17:21:35.8569932Z 4: Expected equality of these values:
2020-02-26T17:21:35.8569999Z 4:   expected_stat.at(stat.data_type)
2020-02-26T17:21:35.8570082Z 4:     Which is: 70.385238647460938
2020-02-26T17:21:35.8570165Z 4:   stat.data
2020-02-26T17:21:35.8570346Z 4:     Which is: 70.330429077148438
2020-02-26T17:21:35.8570463Z 4: /home/runner/work/system_metrics_collector/system_metrics_collector/ros_ws/src/system_metrics_collector/system_metrics_collector/test/system_metrics_collector/test_linux_memory_measurement.cpp:275: Failure
2020-02-26T17:21:35.8570565Z 4: Expected equality of these values:
2020-02-26T17:21:35.8570633Z 4:   expected_stat.at(stat.data_type)
2020-02-26T17:21:35.8570717Z 4:     Which is: 70.385238647460938
2020-02-26T17:21:35.8570800Z 4:   stat.data
2020-02-26T17:21:35.8570879Z 4:     Which is: 70.330429077148438
2020-02-26T17:21:35.8570984Z 4: /home/runner/work/system_metrics_collector/system_metrics_collector/ros_ws/src/system_metrics_collector/system_metrics_collector/test/system_metrics_collector/test_linux_memory_measurement.cpp:275: Failure
2020-02-26T17:21:35.8571128Z 4: Expected equality of these values:
2020-02-26T17:21:35.8571212Z 4:   expected_stat.at(stat.data_type)
2020-02-26T17:21:35.8571298Z 4:     Which is: 70.330429077148438
2020-02-26T17:21:35.8571382Z 4:   stat.data
2020-02-26T17:21:35.8571462Z 4:     Which is: 70.328407287597656
2020-02-26T17:21:35.8571564Z 4: /home/runner/work/system_metrics_collector/system_metrics_collector/ros_ws/src/system_metrics_collector/system_metrics_collector/test/system_metrics_collector/test_linux_memory_measurement.cpp:275: Failure
2020-02-26T17:21:35.8571652Z 4: Expected equality of these values:
2020-02-26T17:21:35.8571735Z 4:   expected_stat.at(stat.data_type)
2020-02-26T17:21:35.8571820Z 4:     Which is: 70.330429077148438
2020-02-26T17:21:35.8571901Z 4:   stat.data
2020-02-26T17:21:35.8571982Z 4:     Which is: 70.33050537109375
2020-02-26T17:21:35.8572074Z 4: /home/runner/work/system_metrics_collector/system_metrics_collector/ros_ws/src/system_metrics_collector/system_metrics_collector/test/system_metrics_collector/test_linux_memory_measurement.cpp:275: Failure
2020-02-26T17:21:35.8572175Z 4: Expected equality of these values:
2020-02-26T17:21:35.8572259Z 4:   expected_stat.at(stat.data_type)
2020-02-26T17:21:35.8572342Z 4:     Which is: 70.330429077148438
2020-02-26T17:21:35.8572424Z 4:   stat.data
2020-02-26T17:21:35.8572505Z 4:     Which is: 70.326309204101562
2020-02-26T17:21:35.8572592Z 4: /home/runner/work/system_metrics_collector/system_metrics_collector/ros_ws/src/system_metrics_collector/system_metrics_collector/test/system_metrics_collector/test_linux_memory_measurement.cpp:275: Failure
2020-02-26T17:21:35.8572692Z 4: Expected equality of these values:
2020-02-26T17:21:35.8572776Z 4:   expected_stat.at(stat.data_type)
2020-02-26T17:21:35.8572859Z 4:     Which is: 1
2020-02-26T17:21:35.8572939Z 4:   stat.data
2020-02-26T17:21:35.8573007Z 4:     Which is: 2
2020-02-26T17:21:35.8573105Z 4: /home/runner/work/system_metrics_collector/system_metrics_collector/ros_ws/src/system_metrics_collector/system_metrics_collector/test/system_metrics_collector/test_linux_memory_measurement.cpp:275: Failure
2020-02-26T17:21:35.8573209Z 4: Expected equality of these values:
2020-02-26T17:21:35.8573294Z 4:   expected_stat.at(stat.data_type)
2020-02-26T17:21:35.8573375Z 4:     Which is: 0
2020-02-26T17:21:35.8573455Z 4:   stat.data
2020-02-26T17:21:35.8573523Z 4:     Which is: 0.0020976180676370859
2020-02-26T17:21:35.8573625Z 4: /home/runner/work/system_metrics_collector/system_metrics_collector/ros_ws/src/system_metrics_collector/system_metrics_collector/test/system_metrics_collector/test_linux_memory_measurement.cpp:275: Failure
2020-02-26T17:21:35.8573724Z 4: Expected equality of these values:
2020-02-26T17:21:35.8573807Z 4:   expected_stat.at(stat.data_type)
2020-02-26T17:21:35.8573889Z 4:     Which is: 70.328407287597656
2020-02-26T17:21:35.8573959Z 4:   stat.data
2020-02-26T17:21:35.8574040Z 4:     Which is: 70.326553344726562
2020-02-26T17:21:35.8574141Z 4: /home/runner/work/system_metrics_collector/system_metrics_collector/ros_ws/src/system_metrics_collector/system_metrics_collector/test/system_metrics_collector/test_linux_memory_measurement.cpp:275: Failure
2020-02-26T17:21:35.8574287Z 4: Expected equality of these values:
2020-02-26T17:21:35.8574380Z 4:   expected_stat.at(stat.data_type)
2020-02-26T17:21:35.8574465Z 4:     Which is: 70.33050537109375
2020-02-26T17:21:35.8574532Z 4:   stat.data
2020-02-26T17:21:35.8574614Z 4:     Which is: 70.326553344726562
2020-02-26T17:21:35.8574717Z 4: /home/runner/work/system_metrics_collector/system_metrics_collector/ros_ws/src/system_metrics_collector/system_metrics_collector/test/system_metrics_collector/test_linux_memory_measurement.cpp:275: Failure
2020-02-26T17:21:35.8574821Z 4: Expected equality of these values:
2020-02-26T17:21:35.8574903Z 4:   expected_stat.at(stat.data_type)
2020-02-26T17:21:35.8575033Z 4:     Which is: 70.326309204101562
2020-02-26T17:21:35.8575101Z 4:   stat.data
2020-02-26T17:21:35.8575369Z 4:     Which is: 70.326553344726562
2020-02-26T17:21:35.8575478Z 4: /home/runner/work/system_metrics_collector/system_metrics_collector/ros_ws/src/system_metrics_collector/system_metrics_collector/test/system_metrics_collector/test_linux_memory_measurement.cpp:275: Failure
2020-02-26T17:21:35.8575579Z 4: Expected equality of these values:
2020-02-26T17:21:35.8575665Z 4:   expected_stat.at(stat.data_type)
2020-02-26T17:21:35.8575734Z 4:     Which is: 2
2020-02-26T17:21:35.8575817Z 4:   stat.data
2020-02-26T17:21:35.8575898Z 4:     Which is: 1
2020-02-26T17:21:35.8576000Z 4: /home/runner/work/system_metrics_collector/system_metrics_collector/ros_ws/src/system_metrics_collector/system_metrics_collector/test/system_metrics_collector/test_linux_memory_measurement.cpp:275: Failure
2020-02-26T17:21:35.8576104Z 4: Expected equality of these values:
2020-02-26T17:21:35.8576188Z 4:   expected_stat.at(stat.data_type)
2020-02-26T17:21:35.8576263Z 4:     Which is: 0.0020976180676370859
2020-02-26T17:21:35.8576348Z 4:   stat.data
2020-02-26T17:21:35.8576430Z 4:     Which is: 0
2020-02-26T17:21:35.8576536Z 4: /home/runner/work/system_metrics_collector/system_metrics_collector/ros_ws/src/system_metrics_collector/system_metrics_collector/test/system_metrics_collector/test_linux_memory_measurement.cpp:371: Failure
2020-02-26T17:21:35.8576636Z 4: Expected equality of these values:
2020-02-26T17:21:35.8576706Z 4:   6
2020-02-26T17:21:35.8576982Z 4:   test_receive_measurements->GetNumReceived()
2020-02-26T17:21:35.8577074Z 4:     Which is: 5
2020-02-26T17:21:35.8577170Z 4: [  FAILED  ] LinuxMemoryMeasurementTestFixture.TestPublishMetricsMessage (766 ms)
2020-02-26T17:21:35.8577443Z 4: [----------] 2 tests from LinuxMemoryMeasurementTestFixture (785 ms total)
2020-02-26T17:21:35.8577523Z 4: 
2020-02-26T17:21:35.8577750Z 4: [----------] Global test environment tear-down
2020-02-26T17:21:35.8577845Z 4: [==========] 3 tests from 2 test cases ran. (785 ms total)
2020-02-26T17:21:35.8577943Z 4: [  PASSED  ] 2 tests.
2020-02-26T17:21:35.8578028Z 4: [  FAILED  ] 1 test, listed below:
2020-02-26T17:21:35.8578123Z 4: [  FAILED  ] LinuxMemoryMeasurementTestFixture.TestPublishMetricsMessage
2020-02-26T17:21:35.8578202Z 4: 
2020-02-26T17:21:35.8578284Z 4:  1 FAILED TEST
2020-02-26T17:21:35.8578686Z 4: -- run_test.py: return code 1
2020-02-26T17:21:35.8579088Z 4: -- run_test.py: inject classname prefix into gtest result file '/home/runner/work/system_metrics_collector/system_metrics_collector/ros_ws/build/system_metrics_collector/test_results/system_metrics_collector/test_linux_memory_measurement_node.gtest.xml'
2020-02-26T17:21:35.8579485Z 4: -- run_test.py: verify result file '/home/runner/work/system_metrics_collector/system_metrics_collector/ros_ws/build/system_metrics_collector/test_results/system_metrics_collector/test_linux_memory_measurement_node.gtest.xml'
```
</details>

## To Reproduce

Run:

```shell
./test_linux_memory_measurement_node --gtest_repeat=99999999
```

Wait until it starts failing. On fast workstations, buildling with ASAN enabled may be necessary to slow-down the execution enough to exhibit the bug.

## System (please complete the following information)
 - OS: Bionic
 - ROS2 Distro: `master`

## Additional context

To make this bug reliable, each step in the test needs to wait until the ROS messages are actually processed. The callback may switch a `std::atomic_bool` or a `std::atomic_flag` for instance. Blocking until the lifecycle API starts reporting the new state may also be another possibility.

If waiting for messages to be processed is not enough, we should explore whether cache coherency could play a role in this bug.

--- a/system_metrics_collector/test/system_metrics_collector/test_constants.hpp
+++ b/system_metrics_collector/test/system_metrics_collector/test_constants.hpp
@@ -26,9 +26,10 @@
  */
 namespace test_constants
 {
-constexpr const std::chrono::milliseconds kTestDuration{250};
 constexpr const std::chrono::milliseconds kMeasurePeriod{50};
 constexpr const std::chrono::milliseconds kPublishPeriod{80};
+constexpr const std::chrono::milliseconds kTestDuration{250};
+constexpr const std::chrono::seconds kSpinTimeout{1};
 
 constexpr const char kProcSampleResolutionTest[] =
   "cpu  57211920 335926 18096939 2526329830 14818556 0 1072048 0 0 0\n";

--- a/system_metrics_collector/test/system_metrics_collector/test_linux_memory_measurement.cpp
+++ b/system_metrics_collector/test/system_metrics_collector/test_linux_memory_measurement.cpp
@@ -15,8 +15,10 @@
 #include <gtest/gtest.h>
 
 #include <array>
+#include <atomic>
 #include <chrono>
 #include <functional>
+#include <future>
 #include <iostream>
 #include <memory>
 #include <mutex>
@@ -29,11 +31,14 @@
 #include "metrics_statistics_msgs/msg/metrics_message.hpp"
 #include "metrics_statistics_msgs/msg/statistic_data_type.hpp"
 
+#include "moving_average_statistics/moving_average.hpp"
+
 #include "system_metrics_collector/constants.hpp"
 #include "system_metrics_collector/linux_memory_measurement_node.hpp"
 #include "system_metrics_collector/utilities.hpp"
 
 #include "test_constants.hpp"
+
 
 using lifecycle_msgs::msg::State;
 using metrics_statistics_msgs::msg::MetricsMessage;
@@ -41,15 +46,19 @@ using metrics_statistics_msgs::msg::StatisticDataPoint;
 using metrics_statistics_msgs::msg::StatisticDataType;
 using moving_average_statistics::StatisticData;
 using system_metrics_collector::ProcessMemInfoLines;
+using moving_average_statistics::MovingAverageStatistics;
+
+using ExpectedStatistics =
+  std::unordered_map<decltype(StatisticDataPoint::data_type), decltype(StatisticDataPoint::data)>;
 
 namespace
 {
 constexpr const char kTestNodeName[] = "test_measure_linux_memory";
 constexpr const char kTestMetricName[] = "system_memory_percent_used";
+constexpr const std::chrono::seconds kPublishTestTimeout{2};
+constexpr const std::chrono::milliseconds kPublishPeriod{150};
 
-constexpr const std::array<const char *, 10> kSamples = {
-  test_constants::kFullSample,
-
+const std::vector<std::string> kSamples = {
   "MemTotal:       16304208 kB\n"
   "MemFree:          845168 kB\n"
   "MemAvailable:    4840176 kB\n",
@@ -59,71 +68,131 @@ constexpr const std::array<const char *, 10> kSamples = {
   "MemAvailable:     239124 kB\n",
 
   "MemTotal:       16304208 kB\n"
-  "MemFree:          821256 kB\n"
-  "MemAvailable:    4828452 kB\n",
-
-  "MemTotal:       16304208 kB\n"
-  "MemFree:          825460 kB\n"
-  "MemAvailable:    4835920 kB\n",
-
-  "MemTotal:       16304208 kB\n"
   "MemFree:          826912 kB\n"
   "MemAvailable:    4837388 kB\n",
-
-  "MemTotal:       16304208 kB\n"
-  "MemFree:          827568 kB\n"
-  "MemAvailable:    4838060 kB\n",
-
-  "MemTotal:       16304208 kB\n"
-  "MemFree:          826792 kB\n"
-  "MemAvailable:    4837376 kB\n",
-
-  "MemTotal:       16304208 kB\n"
-  "MemFree:          827380 kB\n"
-  "MemAvailable:    4838020 kB\n",
-
-  "MemTotal:       16304208 kB\n"
-  "MemFree:          826968 kB\n"
-  "MemAvailable:    4837664 kB\n",
 };
-
 }  // namespace
 
+ExpectedStatistics StatisticDataToExpectedStatistics(const StatisticData & src)
+{
+  ExpectedStatistics expected{};
+  expected[StatisticDataType::STATISTICS_DATA_TYPE_AVERAGE] = src.average;
+  expected[StatisticDataType::STATISTICS_DATA_TYPE_MINIMUM] = src.min;
+  expected[StatisticDataType::STATISTICS_DATA_TYPE_MAXIMUM] = src.max;
+  expected[StatisticDataType::STATISTICS_DATA_TYPE_STDDEV] = src.standard_deviation;
+  expected[StatisticDataType::STATISTICS_DATA_TYPE_SAMPLE_COUNT] = src.sample_count;
+  return expected;
+}
 
-class TestLinuxMemoryMeasurementNode : public system_metrics_collector::LinuxMemoryMeasurementNode
+class PromiseSetter
 {
 public:
-  TestLinuxMemoryMeasurementNode(const std::string & name, const rclcpp::NodeOptions & options)
-  : LinuxMemoryMeasurementNode{name, options},
-    measurement_index_(0) {}
-
-  ~TestLinuxMemoryMeasurementNode() override = default;
-
-  void SetTestString(const std::string & test_string)
+  /**
+   * Reassign the promise member and return it's future.
+   * @return the promise member's future, called upon PeriodicMeasurement
+   */
+  std::shared_future<bool> GetFuture()
   {
-    measurement_index_ = kInvalidIndex;
-    test_string_ = test_string;
+    std::unique_lock<std::mutex> ulock{mutex_};
+    use_future_ = true;
+    promise_ = std::promise<bool>();
+    return promise_.get_future();
   }
 
-  // override to avoid calling methods involved in file i/o
-  double PeriodicMeasurement() override
+protected:
+  void SetPromise()
   {
-    if (measurement_index_ == kInvalidIndex) {
-      return ProcessMemInfoLines(test_string_);
-    } else {
-      EXPECT_GT(kSamples.size(), measurement_index_);
-      return ProcessMemInfoLines(kSamples[measurement_index_++]);
+    std::unique_lock<std::mutex> ulock{mutex_};
+    if (use_future_) {
+      // only set if GetFuture was called
+      promise_.set_value(true);
+      use_future_ = false;    // the promise needs to be reassigned to set again
     }
   }
 
 private:
-  static constexpr int kInvalidIndex = -1;
-  int measurement_index_;
-  std::string test_string_;
+  mutable std::mutex mutex_;
+  std::promise<bool> promise_;
+  bool use_future_{false};
+};
+/**
+ *
+ */
+class TestLinuxMemoryMeasurementNode : public system_metrics_collector::LinuxMemoryMeasurementNode,
+  public PromiseSetter
+{
+public:
+  TestLinuxMemoryMeasurementNode(const std::string & name, const rclcpp::NodeOptions & options)
+  : LinuxMemoryMeasurementNode{name, options} {}
+  ~TestLinuxMemoryMeasurementNode() override = default;
+
+  /**
+   *
+   * @param test_data
+   */
+  void SetTestVector(const std::vector<std::string> & test_data)
+  {
+    test_vector_ = test_data;
+    index_ = 0;
+  }
+
+  /**
+   * Override to avoid calling methods involved in file i/o.
+   *
+   *
+   */
+  double PeriodicMeasurement() override
+  {
+    const auto to_return = ProcessMemInfoLines(test_vector_[index_]);
+    if (++index_ >= test_vector_.size()) {
+      index_ = 0;
+    }
+    PromiseSetter::SetPromise();
+    return to_return;
+  }
+
+private:
+  std::vector<std::string> test_vector_{""};
+  std::atomic<int> index_{0};
 };
 
-constexpr int TestLinuxMemoryMeasurementNode::kInvalidIndex;
+/**
+ * Node which listens for published MetricsMessages.
+ */
+class TestReceiveMemoryMeasurementNode : public rclcpp::Node, public PromiseSetter
+{
+public:
+  explicit TestReceiveMemoryMeasurementNode(const std::string & name)
+  : rclcpp::Node(name)          //, received_(false)
+  {
+    auto callback = [this](MetricsMessage::UniquePtr msg) {this->MetricsMessageCallback(*msg);};
+    subscription_ = create_subscription<MetricsMessage,
+        std::function<void(MetricsMessage::UniquePtr)>>(
+      system_metrics_collector::collector_node_constants::kStatisticsTopicName,
+      0 /*history_depth*/, callback);
+  }
 
+  MetricsMessage GetLastReceivedMessage()
+  {
+    return last_received_message_;   // todo locks
+  }
+
+private:
+  void MetricsMessageCallback(const MetricsMessage & msg)
+  {
+    std::unique_lock<std::mutex> ulock{mutex_};
+    last_received_message_ = msg;
+    PromiseSetter::SetPromise();
+  }
+
+  MetricsMessage last_received_message_;
+  rclcpp::Subscription<MetricsMessage>::SharedPtr subscription_;
+  mutable std::mutex mutex_;
+};
+
+/**
+ *
+ */
 class LinuxMemoryMeasurementTestFixture : public ::testing::Test
 {
 public:
@@ -137,7 +206,7 @@ public:
       test_constants::kMeasurePeriod.count());
     options.append_parameter_override(
       system_metrics_collector::collector_node_constants::kPublishPeriodParam,
-      test_constants::kPublishPeriod.count());
+      kPublishPeriod.count());
 
     test_measure_linux_memory_ = std::make_shared<TestLinuxMemoryMeasurementNode>(
       kTestNodeName, options);
@@ -167,123 +236,6 @@ protected:
   std::shared_ptr<TestLinuxMemoryMeasurementNode> test_measure_linux_memory_;
 };
 
-class TestReceiveMemoryMeasurementNode : public rclcpp::Node
-{
-public:
-  explicit TestReceiveMemoryMeasurementNode(const std::string & name)
-  : rclcpp::Node(name), times_received_(0)
-  {
-    auto callback = [this](MetricsMessage::UniquePtr msg) {this->MetricsMessageCallback(*msg);};
-    subscription_ = create_subscription<MetricsMessage,
-        std::function<void(MetricsMessage::UniquePtr)>>(
-      system_metrics_collector::collector_node_constants::kStatisticsTopicName,
-      10 /*history_depth*/, callback);
-
-    // tools for calculating expected statistics values
-    moving_average_statistics::MovingAverageStatistics stats_calc;
-    StatisticData data;
-
-    // setting expected_stats_[0]
-    // round 1 50 ms: SAMPLES[0] is collected
-    // round 1 80 ms: statistics derived from SAMPLES[0] is published
-    stats_calc.Reset();
-    stats_calc.AddMeasurement(ProcessMemInfoLines(kSamples[0]));
-    data = stats_calc.GetStatistics();
-    StatisticDataToExpectedStatistics(data, expected_stats_[0]);
-
-    // setting expected_stats_[1]
-    // round 1 100 ms: SAMPLES[1] is collected
-    // round 1 150 ms: SAMPLES[2] is collected
-    // round 1 160 ms: statistics derived from SAMPLES[1 & 2] is published
-    stats_calc.Reset();
-    stats_calc.AddMeasurement(ProcessMemInfoLines(kSamples[1]));
-    stats_calc.AddMeasurement(ProcessMemInfoLines(kSamples[2]));
-    data = stats_calc.GetStatistics();
-    StatisticDataToExpectedStatistics(data, expected_stats_[1]);
-
-    // setting expected_stats_[2]
-    // round 1 200 ms: SAMPLES[3] is collected
-    // round 1 240 ms: statistics derived from SAMPLES[3] is published
-    stats_calc.Reset();
-    stats_calc.AddMeasurement(ProcessMemInfoLines(kSamples[3]));
-    data = stats_calc.GetStatistics();
-    StatisticDataToExpectedStatistics(data, expected_stats_[2]);
-
-    // setting expected_stats_[3]
-    // round 2 50 ms: SAMPLES[5] is collected
-    // round 2 80 ms: statistics derived from SAMPLES[5] is published
-    stats_calc.Reset();
-    stats_calc.AddMeasurement(ProcessMemInfoLines(kSamples[5]));
-    data = stats_calc.GetStatistics();
-    StatisticDataToExpectedStatistics(data, expected_stats_[3]);
-
-    // setting expected_stats_[4]
-    // round 2 100 ms: SAMPLES[6] is collected
-    // round 2 150 ms: SAMPLES[7] is collected
-    // round 2 160 ms: statistics derived from SAMPLES[6 & 7] is published
-    stats_calc.Reset();
-    stats_calc.AddMeasurement(ProcessMemInfoLines(kSamples[6]));
-    stats_calc.AddMeasurement(ProcessMemInfoLines(kSamples[7]));
-    data = stats_calc.GetStatistics();
-    StatisticDataToExpectedStatistics(data, expected_stats_[4]);
-
-    // setting expected_stats_[5]
-    // round 2 200 ms: SAMPLES[8] is collected
-    // round 2 240 ms: statistics derived from SAMPLES[8] is published
-    stats_calc.Reset();
-    stats_calc.AddMeasurement(ProcessMemInfoLines(kSamples[8]));
-    data = stats_calc.GetStatistics();
-    StatisticDataToExpectedStatistics(data, expected_stats_[5]);
-  }
-
-  int GetNumReceived() const
-  {
-    return times_received_;
-  }
-
-private:
-  using ExpectedStatistics =
-    std::unordered_map<decltype(StatisticDataPoint::data_type), decltype(StatisticDataPoint::data)>;
-
-  void StatisticDataToExpectedStatistics(const StatisticData & src, ExpectedStatistics & dst)
-  {
-    dst[StatisticDataType::STATISTICS_DATA_TYPE_AVERAGE] = src.average;
-    dst[StatisticDataType::STATISTICS_DATA_TYPE_MINIMUM] = src.min;
-    dst[StatisticDataType::STATISTICS_DATA_TYPE_MAXIMUM] = src.max;
-    dst[StatisticDataType::STATISTICS_DATA_TYPE_STDDEV] = src.standard_deviation;
-    dst[StatisticDataType::STATISTICS_DATA_TYPE_SAMPLE_COUNT] = src.sample_count;
-  }
-
-  void MetricsMessageCallback(const MetricsMessage & msg) const
-  {
-    ASSERT_GT(expected_stats_.size(), times_received_);
-
-    // check source names
-    EXPECT_EQ(kTestNodeName, msg.measurement_source_name);
-    EXPECT_EQ(kTestMetricName, msg.metrics_source);
-    EXPECT_EQ(system_metrics_collector::collector_node_constants::kPercentUnitName, msg.unit);
-
-    // check measurements
-    const ExpectedStatistics & expected_stat = expected_stats_[times_received_];
-    EXPECT_EQ(expected_stat.size(), msg.statistics.size());
-
-    for (const StatisticDataPoint & stat : msg.statistics) {
-      EXPECT_GT(expected_stat.count(stat.data_type), 0);
-      if (std::isnan(expected_stat.at(stat.data_type))) {
-        EXPECT_TRUE(std::isnan(stat.data));
-      } else {
-        EXPECT_DOUBLE_EQ(expected_stat.at(stat.data_type), stat.data);
-      }
-    }
-
-    ++times_received_;
-  }
-
-  rclcpp::Subscription<MetricsMessage>::SharedPtr subscription_;
-  std::array<ExpectedStatistics, 6> expected_stats_;
-  mutable int times_received_;
-};
-
 TEST(LinuxMemoryMeasurementTest, TestReadInvalidFile)
 {
   const auto s = system_metrics_collector::ReadFileToString("this_will_fail.txt");
@@ -291,16 +243,16 @@ TEST(LinuxMemoryMeasurementTest, TestReadInvalidFile)
 }
 
 TEST_F(LinuxMemoryMeasurementTestFixture, testManualMeasurement) {
-  test_measure_linux_memory_->SetTestString("");
   double mem_used_percentage = test_measure_linux_memory_->PeriodicMeasurement();
   ASSERT_TRUE(std::isnan(mem_used_percentage));
 
-  test_measure_linux_memory_->SetTestString(test_constants::kFullSample);
+  const auto tv_valid = std::vector<std::string>{test_constants::kFullSample};
+  test_measure_linux_memory_->SetTestVector(tv_valid);
   mem_used_percentage = test_measure_linux_memory_->PeriodicMeasurement();
   ASSERT_DOUBLE_EQ(test_constants::kMemoryUsedPercentage, mem_used_percentage);
 }
 
-TEST_F(LinuxMemoryMeasurementTestFixture, TestPublishMetricsMessage)
+TEST_F(LinuxMemoryMeasurementTestFixture, TestPeriodicMeasurement)
 {
   ASSERT_NE(test_measure_linux_memory_, nullptr);
   ASSERT_FALSE(test_measure_linux_memory_->IsStarted());
@@ -308,37 +260,34 @@ TEST_F(LinuxMemoryMeasurementTestFixture, TestPublishMetricsMessage)
     State::PRIMARY_STATE_UNCONFIGURED,
     test_measure_linux_memory_->get_current_state().id());
 
-  auto test_receive_measurements = std::make_shared<TestReceiveMemoryMeasurementNode>(
-    "test_receive_measurements");
+  // set with a single valid sample
+  const auto tv_valid = std::vector<std::string>{test_constants::kFullSample};
+  test_measure_linux_memory_->SetTestVector(tv_valid);
+
   std::promise<bool> empty_promise;
   std::shared_future<bool> dummy_future = empty_promise.get_future();
+
   rclcpp::executors::SingleThreadedExecutor ex;
   ex.add_node(test_measure_linux_memory_->get_node_base_interface());
-  ex.add_node(test_receive_measurements->get_node_base_interface());
 
-  //
-  // spin the node with it started
-  //
+  // configure the node manually (lifecycle transition)
   test_measure_linux_memory_->configure();
   ASSERT_EQ(State::PRIMARY_STATE_INACTIVE, test_measure_linux_memory_->get_current_state().id());
   ASSERT_FALSE(test_measure_linux_memory_->IsStarted());
 
+  // activate the node manually (lifecycle transition): this allows collection and data publication
   test_measure_linux_memory_->activate();
   ASSERT_EQ(State::PRIMARY_STATE_ACTIVE, test_measure_linux_memory_->get_current_state().id());
   ASSERT_TRUE(test_measure_linux_memory_->IsStarted());
 
-  ex.spin_until_future_complete(dummy_future, test_constants::kTestDuration);
-  EXPECT_EQ(3, test_receive_measurements->GetNumReceived());
-  // expectation is:
-  // 50 ms: SAMPLES[0] is collected
-  // 80 ms: statistics derived from SAMPLES[0] is published. statistics are cleared
-  // 100 ms: SAMPLES[1] is collected
-  // 150 ms: SAMPLES[2] is collected
-  // 160 ms: statistics derived from SAMPLES[1 & 2] is published. statistics are cleared
-  // 200 ms: SAMPLES[3] is collected
-  // 240 ms: statistics derived from SAMPLES[3] is published. statistics are cleared
-  // 250 ms: SAMPLES[4] is collected. last GetStatisticsResults() is of SAMPLES[4]
-  StatisticData data = test_measure_linux_memory_->GetStatisticsResults();
+  //
+  // spin the node while activated and use the node's future to halt after the first measurement
+  //
+  ex.spin_until_future_complete(
+    test_measure_linux_memory_->GetFuture(), test_constants::kSpinTimeout);
+
+  // expect that a single measurement will be made
+  auto data = test_measure_linux_memory_->GetStatisticsResults();
   EXPECT_EQ(1, data.sample_count);
 
   //
@@ -348,8 +297,8 @@ TEST_F(LinuxMemoryMeasurementTestFixture, TestPublishMetricsMessage)
   ASSERT_EQ(State::PRIMARY_STATE_INACTIVE, test_measure_linux_memory_->get_current_state().id());
   ASSERT_FALSE(test_measure_linux_memory_->IsStarted());
 
-  ex.spin_until_future_complete(dummy_future, test_constants::kTestDuration);
-  EXPECT_EQ(3, test_receive_measurements->GetNumReceived());
+  // use the dummy future as the test_measure_linux_memory_ promise won't be set
+  ex.spin_until_future_complete(dummy_future, test_constants::kSpinTimeout);
   // expectation is:
   // upon calling stop, samples are cleared, so GetStatisticsResults() would be NaNs
   // no MetricsMessages are published
@@ -360,24 +309,101 @@ TEST_F(LinuxMemoryMeasurementTestFixture, TestPublishMetricsMessage)
   EXPECT_TRUE(std::isnan(data.standard_deviation));
   EXPECT_EQ(0, data.sample_count);
 
-  //
-  // spin the node with it restarted
-  //
+  // reactivate the node
   test_measure_linux_memory_->activate();
   ASSERT_EQ(State::PRIMARY_STATE_ACTIVE, test_measure_linux_memory_->get_current_state().id());
 
+  //
+  // spin the reactivated node and use the node's future to halt after the first measurement
+  //
   ASSERT_TRUE(test_measure_linux_memory_->IsStarted());
-  ex.spin_until_future_complete(dummy_future, test_constants::kTestDuration);
-  EXPECT_EQ(6, test_receive_measurements->GetNumReceived());
-  // expectation is:
-  // 50 ms: SAMPLES[5] is collected
-  // 80 ms: statistics derived from SAMPLES[5] is published. statistics are cleared
-  // 100 ms: SAMPLES[6] is collected
-  // 150 ms: SAMPLES[7] is collected
-  // 160 ms: statistics derived from SAMPLES[6 & 7] is published. statistics are cleared
-  // 200 ms: SAMPLES[8] is collected
-  // 240 ms: statistics derived from SAMPLES[8] is published. statistics are cleared
-  // 250 ms: SAMPLES[9] is collected. last GetStatisticsResults() is of SAMPLES[9]
+  ex.spin_until_future_complete(
+    test_measure_linux_memory_->GetFuture(), test_constants::kSpinTimeout);
+
   data = test_measure_linux_memory_->GetStatisticsResults();
+
   EXPECT_EQ(1, data.sample_count);
+  EXPECT_DOUBLE_EQ(test_constants::kMemoryUsedPercentage, data.average);
+  EXPECT_DOUBLE_EQ(test_constants::kMemoryUsedPercentage, data.min);
+  EXPECT_DOUBLE_EQ(test_constants::kMemoryUsedPercentage, data.max);
+  EXPECT_DOUBLE_EQ(0.0, data.standard_deviation);  // only one sample so 0
+}
+
+TEST_F(LinuxMemoryMeasurementTestFixture, TestPublishMessage)
+{
+  ASSERT_NE(test_measure_linux_memory_, nullptr);
+  ASSERT_FALSE(test_measure_linux_memory_->IsStarted());
+  ASSERT_EQ(
+    State::PRIMARY_STATE_UNCONFIGURED,
+    test_measure_linux_memory_->get_current_state().id());
+
+  // set with a single valid sample
+  const auto tv_valid = std::vector<std::string>{test_constants::kFullSample};
+  test_measure_linux_memory_->SetTestVector(kSamples);
+
+  auto test_receive_measurements = std::make_shared<TestReceiveMemoryMeasurementNode>(
+    "test_receive_measurements");
+
+  rclcpp::executors::SingleThreadedExecutor ex;
+  ex.add_node(test_measure_linux_memory_->get_node_base_interface());
+  ex.add_node(test_receive_measurements);
+
+  // configure the node manually (lifecycle transition)
+  test_measure_linux_memory_->configure();
+  ASSERT_EQ(State::PRIMARY_STATE_INACTIVE, test_measure_linux_memory_->get_current_state().id());
+  ASSERT_FALSE(test_measure_linux_memory_->IsStarted());
+
+  // activate the node manually (lifecycle transition): this allows collection and data publication
+  test_measure_linux_memory_->activate();
+  ASSERT_EQ(State::PRIMARY_STATE_ACTIVE, test_measure_linux_memory_->get_current_state().id());
+  ASSERT_TRUE(test_measure_linux_memory_->IsStarted());
+
+  //
+  // spin the node while activated and use the node's future to halt
+  // after the first published message
+  //
+  ex.spin_until_future_complete(
+    test_receive_measurements->GetFuture(), kPublishTestTimeout);
+
+  MovingAverageStatistics expected_moving_average;
+  for (const std::string sample : kSamples) {
+    const auto d = ProcessMemInfoLines(sample);
+    expected_moving_average.AddMeasurement(d);
+  }
+
+  const auto expected_stats = StatisticDataToExpectedStatistics(
+    expected_moving_average.GetStatistics());
+
+  // check expected received message
+  const auto received_message = test_receive_measurements->GetLastReceivedMessage();
+  EXPECT_EQ(kTestNodeName, received_message.measurement_source_name);
+  EXPECT_EQ(kTestMetricName, received_message.metrics_source);
+  EXPECT_EQ(
+    system_metrics_collector::collector_node_constants::kPercentUnitName,
+    received_message.unit);
+
+  for (const StatisticDataPoint & stats_point : received_message.statistics) {
+    const auto type = stats_point.data_type;
+    switch (type) {
+      case StatisticDataType::STATISTICS_DATA_TYPE_SAMPLE_COUNT:
+        EXPECT_DOUBLE_EQ(expected_stats.at(type), stats_point.data) << "unexpected sample count";
+        break;
+      case StatisticDataType::STATISTICS_DATA_TYPE_AVERAGE:
+        EXPECT_DOUBLE_EQ(expected_stats.at(type), stats_point.data) << "unexpected average";
+        break;
+      case StatisticDataType::STATISTICS_DATA_TYPE_MINIMUM:
+        EXPECT_DOUBLE_EQ(
+          expected_stats.at(type), stats_point.data) << "unexpected min";
+        break;
+      case StatisticDataType::STATISTICS_DATA_TYPE_MAXIMUM:
+        EXPECT_DOUBLE_EQ(
+          expected_stats.at(type), stats_point.data) << "unexpected max";
+        break;
+      case StatisticDataType::STATISTICS_DATA_TYPE_STDDEV:
+        EXPECT_DOUBLE_EQ(expected_stats.at(type), stats_point.data) << "unexpected stddev";
+        break;
+      default:
+        FAIL() << "received unknown statistics type: " << std::to_string(type);
+    }
+  }
 }

--- a/system_metrics_collector/test/system_metrics_collector/test_linux_memory_measurement.cpp
+++ b/system_metrics_collector/test/system_metrics_collector/test_linux_memory_measurement.cpp
@@ -73,6 +73,13 @@ const std::vector<std::string> kSamples = {
 };
 }  // namespace
 
+/**
+ * Convert input StatisticData to a map of StatisticDataPoint::data_type ->
+ * StatisticDataPoint::data.
+ *
+ * @param src input data to convert
+ * @return output ExpectedStatistics to use for testing.
+ */
 ExpectedStatistics StatisticDataToExpectedStatistics(const StatisticData & src)
 {
   ExpectedStatistics expected{};
@@ -83,12 +90,52 @@ ExpectedStatistics StatisticDataToExpectedStatistics(const StatisticData & src)
   expected[StatisticDataType::STATISTICS_DATA_TYPE_SAMPLE_COUNT] = src.sample_count;
   return expected;
 }
+/**
+ * Check statistic equality. Fails for any unknown statistic type
+ *
+ * @param expected_stats expected data
+ * @param actual data from a received ROS2 message
+ */
+void ExpectedStatisticEquals(
+  const ExpectedStatistics & expected_stats,
+  const metrics_statistics_msgs::msg::MetricsMessage & actual)
+{
+  for (const StatisticDataPoint & stats_point : actual.statistics) {
+    const auto type = stats_point.data_type;
+    switch (type) {
+      case StatisticDataType::STATISTICS_DATA_TYPE_SAMPLE_COUNT:
+        EXPECT_DOUBLE_EQ(expected_stats.at(type), stats_point.data) << "unexpected sample count";
+        break;
+      case StatisticDataType::STATISTICS_DATA_TYPE_AVERAGE:
+        EXPECT_DOUBLE_EQ(expected_stats.at(type), stats_point.data) << "unexpected average";
+        break;
+      case StatisticDataType::STATISTICS_DATA_TYPE_MINIMUM:
+        EXPECT_DOUBLE_EQ(
+          expected_stats.at(type), stats_point.data) << "unexpected min";
+        break;
+      case StatisticDataType::STATISTICS_DATA_TYPE_MAXIMUM:
+        EXPECT_DOUBLE_EQ(
+          expected_stats.at(type), stats_point.data) << "unexpected max";
+        break;
+      case StatisticDataType::STATISTICS_DATA_TYPE_STDDEV:
+        EXPECT_DOUBLE_EQ(expected_stats.at(type), stats_point.data) << "unexpected stddev";
+        break;
+      default:
+        FAIL() << "received unknown statistics type: " << std::to_string(type);
+    }
+  }
+}
 
+/**
+ * Provide an interface to wait for a promise to be satisfied via its future.
+ */
 class PromiseSetter
 {
 public:
   /**
-   * Reassign the promise member and return it's future.
+   * Reassign the promise member and return it's future. Acquires a mutex in order
+   * to mutate member variables.
+   *
    * @return the promise member's future, called upon PeriodicMeasurement
    */
   std::shared_future<bool> GetFuture()
@@ -100,6 +147,10 @@ public:
   }
 
 protected:
+  /**
+   * Set the promise to true, which signals the corresponding future. Acquires a mutex and sets
+   * the promise to true iff GetFuture was invoked before this.
+   */
   void SetPromise()
   {
     std::unique_lock<std::mutex> ulock{mutex_};
@@ -115,8 +166,10 @@ private:
   std::promise<bool> promise_;
   bool use_future_{false};
 };
+
 /**
- *
+ * Mock class which iterates through fake data (provided externally) in order to
+ * yield a periodic measurement.
  */
 class TestLinuxMemoryMeasurementNode : public system_metrics_collector::LinuxMemoryMeasurementNode,
   public PromiseSetter
@@ -127,6 +180,7 @@ public:
   ~TestLinuxMemoryMeasurementNode() override = default;
 
   /**
+   * Set data used by PeriodicMeasurement.
    *
    * @param test_data
    */
@@ -137,9 +191,8 @@ public:
   }
 
   /**
-   * Override to avoid calling methods involved in file i/o.
-   *
-   *
+   * Override to avoid calling methods involved in file i/o. Returns the measurement
+   * of the latest entry of test_vector_.
    */
   double PeriodicMeasurement() override
   {
@@ -152,18 +205,20 @@ public:
   }
 
 private:
-  std::vector<std::string> test_vector_{""};
+  std::vector<std::string> test_vector_{""};  // defaults to invalid data
   std::atomic<int> index_{0};
 };
 
 /**
- * Node which listens for published MetricsMessages.
+ * Node which listens for published MetricsMessages. This uses the PromiseSetter API
+ * in order to signal, via a future, that rclcpp should stop spinning upon
+ * message handling.
  */
 class TestReceiveMemoryMeasurementNode : public rclcpp::Node, public PromiseSetter
 {
 public:
   explicit TestReceiveMemoryMeasurementNode(const std::string & name)
-  : rclcpp::Node(name)          //, received_(false)
+  : rclcpp::Node(name)
   {
     auto callback = [this](MetricsMessage::UniquePtr msg) {this->MetricsMessageCallback(*msg);};
     subscription_ = create_subscription<MetricsMessage,
@@ -172,12 +227,22 @@ public:
       0 /*history_depth*/, callback);
   }
 
+  /**
+   * Acquires a mutex in order to get the last message received member.
+   * @return the last message received
+   */
   MetricsMessage GetLastReceivedMessage()
   {
-    return last_received_message_;   // todo locks
+    std::unique_lock<std::mutex> ulock{mutex_};
+    return last_received_message_;
   }
 
 private:
+  /**
+   * Subscriber callback. Aquires a mutex to set the last message received and
+   * sets the promise to true
+   * @param msg
+   */
   void MetricsMessageCallback(const MetricsMessage & msg)
   {
     std::unique_lock<std::mutex> ulock{mutex_};
@@ -190,9 +255,7 @@ private:
   mutable std::mutex mutex_;
 };
 
-/**
- *
- */
+
 class LinuxMemoryMeasurementTestFixture : public ::testing::Test
 {
 public:
@@ -252,6 +315,9 @@ TEST_F(LinuxMemoryMeasurementTestFixture, testManualMeasurement) {
   ASSERT_DOUBLE_EQ(test_constants::kMemoryUsedPercentage, mem_used_percentage);
 }
 
+/**
+ * Test the lifecycle and check that measurements can be made.
+ */
 TEST_F(LinuxMemoryMeasurementTestFixture, TestPeriodicMeasurement)
 {
   ASSERT_NE(test_measure_linux_memory_, nullptr);
@@ -301,7 +367,6 @@ TEST_F(LinuxMemoryMeasurementTestFixture, TestPeriodicMeasurement)
   ex.spin_until_future_complete(dummy_future, test_constants::kSpinTimeout);
   // expectation is:
   // upon calling stop, samples are cleared, so GetStatisticsResults() would be NaNs
-  // no MetricsMessages are published
   data = test_measure_linux_memory_->GetStatisticsResults();
   EXPECT_TRUE(std::isnan(data.average));
   EXPECT_TRUE(std::isnan(data.min));
@@ -329,6 +394,10 @@ TEST_F(LinuxMemoryMeasurementTestFixture, TestPeriodicMeasurement)
   EXPECT_DOUBLE_EQ(0.0, data.standard_deviation);  // only one sample so 0
 }
 
+/**
+ * Test receipt of a single published message. Expectation is 3 periodic measurements
+ * will be made and one message sent.
+ */
 TEST_F(LinuxMemoryMeasurementTestFixture, TestPublishMessage)
 {
   ASSERT_NE(test_measure_linux_memory_, nullptr);
@@ -337,11 +406,9 @@ TEST_F(LinuxMemoryMeasurementTestFixture, TestPublishMessage)
     State::PRIMARY_STATE_UNCONFIGURED,
     test_measure_linux_memory_->get_current_state().id());
 
-  // set with a single valid sample
-  const auto tv_valid = std::vector<std::string>{test_constants::kFullSample};
   test_measure_linux_memory_->SetTestVector(kSamples);
 
-  auto test_receive_measurements = std::make_shared<TestReceiveMemoryMeasurementNode>(
+  const auto test_receive_measurements = std::make_shared<TestReceiveMemoryMeasurementNode>(
     "test_receive_measurements");
 
   rclcpp::executors::SingleThreadedExecutor ex;
@@ -365,8 +432,9 @@ TEST_F(LinuxMemoryMeasurementTestFixture, TestPublishMessage)
   ex.spin_until_future_complete(
     test_receive_measurements->GetFuture(), kPublishTestTimeout);
 
+  // generate expected data
   MovingAverageStatistics expected_moving_average;
-  for (const std::string sample : kSamples) {
+  for (const std::string & sample : kSamples) {
     const auto d = ProcessMemInfoLines(sample);
     expected_moving_average.AddMeasurement(d);
   }
@@ -376,34 +444,12 @@ TEST_F(LinuxMemoryMeasurementTestFixture, TestPublishMessage)
 
   // check expected received message
   const auto received_message = test_receive_measurements->GetLastReceivedMessage();
+
   EXPECT_EQ(kTestNodeName, received_message.measurement_source_name);
   EXPECT_EQ(kTestMetricName, received_message.metrics_source);
   EXPECT_EQ(
     system_metrics_collector::collector_node_constants::kPercentUnitName,
     received_message.unit);
 
-  for (const StatisticDataPoint & stats_point : received_message.statistics) {
-    const auto type = stats_point.data_type;
-    switch (type) {
-      case StatisticDataType::STATISTICS_DATA_TYPE_SAMPLE_COUNT:
-        EXPECT_DOUBLE_EQ(expected_stats.at(type), stats_point.data) << "unexpected sample count";
-        break;
-      case StatisticDataType::STATISTICS_DATA_TYPE_AVERAGE:
-        EXPECT_DOUBLE_EQ(expected_stats.at(type), stats_point.data) << "unexpected average";
-        break;
-      case StatisticDataType::STATISTICS_DATA_TYPE_MINIMUM:
-        EXPECT_DOUBLE_EQ(
-          expected_stats.at(type), stats_point.data) << "unexpected min";
-        break;
-      case StatisticDataType::STATISTICS_DATA_TYPE_MAXIMUM:
-        EXPECT_DOUBLE_EQ(
-          expected_stats.at(type), stats_point.data) << "unexpected max";
-        break;
-      case StatisticDataType::STATISTICS_DATA_TYPE_STDDEV:
-        EXPECT_DOUBLE_EQ(expected_stats.at(type), stats_point.data) << "unexpected stddev";
-        break;
-      default:
-        FAIL() << "received unknown statistics type: " << std::to_string(type);
-    }
-  }
+  ExpectedStatisticEquals(expected_stats, received_message);
 }

